source/geod24/concurrency.d

```diff
@@ -1063,622 +986,713 @@ protected:
                  !notified && !period.isNegative;
                  period = limit - MonoTime.currTime)
             {
-                FiberScheduler.yield();
+                this.outer.yield();
             }
+
             return notified;
         }
 
-        override void notify() nothrow
+
+        /***********************************************************************
+
+            Notifies one waiter.
+
+        ***********************************************************************/
+
+        public override void notify () nothrow
         {
             notified = true;
-            FiberScheduler.yield();
+            switchContext();
         }
 
-        override void notifyAll() nothrow
+
+        /***********************************************************************
+
+            Notifies all waiters.
+
+        ***********************************************************************/
+
+        public override void notifyAll () nothrow
         {
             notified = true;
-            FiberScheduler.yield();
+            switchContext();
+        }
+
+
+        /***********************************************************************
+
+            switch Fiber Context
+
+        ***********************************************************************/
+
+        private void switchContext() nothrow
+        {
+            if (mutex_nothrow) mutex_nothrow.unlock_nothrow();
+            scope (exit)
+                if (mutex_nothrow)
+                    mutex_nothrow.lock_nothrow();
+            this.outer.yield();
         }
 
         private bool notified;
     }
 
-private:
-    void dispatch()
+    private void dispatch ()
     {
         import std.algorithm.mutation : remove;
 
+        this.mutex.lock_nothrow();
+        scope(exit) this.mutex.unlock_nothrow();
+
+        if (this.dispatching) return;
+
+        this.dispatching = true;
         while (m_fibers.length > 0)
         {
-            auto t = m_fibers[m_pos].call(Fiber.Rethrow.no);
-            if (t !is null && !(cast(OwnerTerminated) t))
-            {
-                throw t;
-            }
-            if (m_fibers[m_pos].state == Fiber.State.TERM)
-            {
-                if (m_pos >= (m_fibers = remove(m_fibers, m_pos)).length)
+            try {
+                auto t = m_fibers[m_pos].call(Fiber.Rethrow.no);
+                if (t !is null && !(cast(ChannelClosed) t) && !(cast(WaitingClosed) t))
+                {
+                    throw t;
+                }
+                if (m_fibers[m_pos].state == Fiber.State.TERM)
+                {
+                    if (m_pos >= (m_fibers = remove(m_fibers, m_pos)).length)
+                        m_pos = 0;
+                }
+                else if (m_pos++ >= m_fibers.length - 1)
+                {
                     m_pos = 0;
+                }
+            } catch (Exception) {
```

Comment:
Why is the Exception ignored here ?

Reply:
```
                if (t !is null && !(cast(ChannelClosed) t) && !(cast(WaitingClosed) t))
                {
                    throw t;
                }
```
`ChannelClosed` and `WaitingClosed` are used as a means to remove the fiber from the array after it is terminated. Another error throws back here.

Reply:
That does not seem correct. In the original code, if `OwnerTerminated` is received, the code just continues and behaves as if nothing happens. Any other `Exception` would have been thrown by the user and thus needs to be re-thrown to user code.

So what the new code is doing is : Ignoring `ChannelClosed` and `WaitingClosed` (alright, I can understand why), but then any user code is simply swallowed and ignored.